v2/pkg/engine/datasource/grpc_datasource/rfcs/field-resolvers.md

```diff
@@ -0,0 +1,527 @@
+
+Consider a GraphQL schema that demonstrates field resolvers with arguments:
+
+```graphql
+type Query {
+  user(id: ID!): User
+  users(limit: Int, offset: Int): [User!]!
+}
+
+type User {
+  id: ID!
+  name: String!
+  email: String!
+  posts(limit: Int = 10, status: PostStatus, orderBy: PostOrderBy): [Post!]!
+}
+
+type Post {
+  id: ID!
+  title: String!
+  content(format: ContentFormat = MARKDOWN): String!
+  comments(limit: Int = 5, orderBy: CommentOrder): [Comment!]!
+  likes(count: Boolean = false): LikesResult!
+}
+
+enum PostStatus {
+  DRAFT
+  PUBLISHED
+  ARCHIVED
+}
+enum PostOrderBy {
+  CREATED_AT
+  TITLE
+  POPULARITY
+}
+enum ContentFormat {
+  MARKDOWN
+  HTML
+  PLAIN
+}
+enum CommentOrder {
+  NEWEST
+  OLDEST
+  POPULAR
+}
+```
+
+In this schema, field resolvers handle arguments to:
+
+- `posts(limit, status, orderBy)` - Filter and paginate user's posts based on arguments
+- `content(format)` - Return post content in different formats
+- `comments(limit, orderBy)` - Control comment pagination and sorting
+- `likes(count)` - Return either like count or full like data based on the boolean argument
+
+Each field resolver receives these arguments and can use them to customize the data fetching logic, API calls, or computations performed for that specific field.
+
+## Field Resolvers in gRPC
+
+There is no specific support for field resolvers in Protobuf. The design focuses on explicitness.
+In order to use field resolvers, we need to create a concept which provides the arguments to the gRPC request and allows the user to provide the response in a way that the engine can interpret the result.
+
+## Concept
+
+Let's focus on the previous example and see how we can implement field resolvers for the `posts` field.
+
+A typical Protobuf schema for the Query `user` would look like this:
+
+```protobuf
+service UserService {
+  rpc QueryUser(QueryUserRequest) returns (QueryUserResponse);
+}
+
+message QueryUserRequest {
+  string id = 1;
+}
+
+message QueryUserResponse {
+  User user = 1;
+}
+```
+
+We provide the `id` parameter defined on the root field `user` in the GraphQL schema.
+Now let's imagine we have the following GraphQL request:
+
+```graphql
+query {
+  user(id: "123") {
+    id
+    name
+    publishedPosts: posts(limit: 10, status: PUBLISHED, orderBy: CREATED_AT) {
+      title
+    }
+    draftPosts: posts(limit: 10, status: DRAFT, orderBy: CREATED_AT) {
+      title
+    }
+  }
+}
+```
+
+We want to fetch the posts for the user with the id `123`, and also compute the value for the `posts` field.
+
+## Approach 1: Generate Lookup rpc calls for each field with a resolver.
+
+```protobuf
+service UserService {
+  rpc QueryUser(QueryUserRequest) returns (QueryUserResponse);
+  rpc ResolveUserPosts(ResolveUserPostsRequest) returns (ResolveUserPostsResponse);
+}
+
+message QueryUserRequest {
+  string id = 1;
+}
+
+message QueryUserResponse {
+  User user = 1;
+}
+
+message ResolveUserPostsInput {
+    string id = 1;
```

Comment:
I think you need to namespace keys and args, otherwise you'll have naming conflicts.